File: server/routes/pandit.js
const router = require('express').Router();
const path = require('path');
const { db } = require('../db');
const { requireRole } = require('../auth');
const B = require('../services/bookings');
const S = require('../lib/serialize');
const upload = require('../lib/upload');
const { verifyOtp } = require('./auth');
const { v, bad, conflict, j, today, rid } = require('../lib/util');
const P = require('../../shared/pricing');

const pujaIds = () => db.prepare('SELECT id FROM pujas').all().map((r) => r.id);
const list = (x) => (Array.isArray(x) ? x : String(x || '').split(',')).map((s) => String(s).trim()).filter(Boolean);
const CITIES = ['Delhi NCR', 'Mumbai', 'Bengaluru', 'Pune', 'Jaipur', 'Lucknow', 'Varanasi', 'Ahmedabad', 'Chennai', 'Hyderabad'];

/* Public: registration with KYC documents (multipart). The mobile number must be verified by OTP. */
router.post('/register', upload.kyc.fields([{ name: 'idDoc', maxCount: 1 }, { name: 'cert', maxCount: 1 }, { name: 'photo', maxCount: 1 }]), upload.verifyMagic(), (req, res) => {
  const b = req.body, mobile = v.mobile(b.mobile);
  verifyOtp(mobile, b.otp);
  const name = v.str(b.name, 'Name', { max: 80 });
  const spec = list(b.spec).filter((s) => pujaIds().includes(s)), langs = list(b.langs).slice(0, 10);
  if (!spec.length) throw bad('Select at least one puja');
  if (!langs.length) throw bad('Select at least one language');
  if (!req.files || !req.files.idDoc) throw bad('Attach an ID document');
  if (db.prepare('SELECT 1 FROM users WHERE mobile=?').get(mobile)) throw conflict('This mobile number is already registered');
  const uid = 'pu' + rid(3), pid = 'p' + rid(3);
  const files = {}; Object.entries(req.files).forEach(([k, f]) => { files[k] = f[0].filename; });
  db.prepare("INSERT INTO users(id,role,name,mobile,joined,created_at) VALUES(?,'pandit',?,?,?,?)").run(uid, name, mobile, today(), Date.now());
  db.prepare("INSERT INTO pandits(id,user_id,name,city,exp,langs,spec,pf,bio,color,status,mobile,kyc) VALUES(?,?,?,?,?,?,?,0.9,'New pandit applicant.','#555555','pending',?,?)")
    .run(pid, uid, name, CITIES.includes(b.city) ? b.city : 'Delhi NCR', v.int(b.exp || 0, 'Experience', { min: 0, max: 70 }), JSON.stringify(langs), JSON.stringify(spec), mobile, JSON.stringify({ idType: String(b.idType || '').slice(0, 30), files }));
  res.status(201).json({ ok: true });
});

router.use(requireRole('pandit'));
const pid = (req) => req.auth.pid;

router.post('/bookings/:id/:action(accept|reject|start)', (req, res) => {
  const p = db.prepare('SELECT status FROM pandits WHERE id=?').get(pid(req));
  if (p.status !== 'verified') throw bad('Your KYC is not verified yet');
  res.json({ booking: S.booking(B.panditAct(pid(req), req.params.id, req.params.action)) });
});
router.post('/bookings/:id/complete', upload.media.array('media', 8), upload.verifyMagic(), (req, res) => {
  const urls = (req.files || []).map((f) => '/media/' + f.filename);
  res.json({ booking: S.booking(B.panditAct(pid(req), req.params.id, 'complete', urls)) });
});
router.post('/availability', (req, res) => {
  const date = v.date(req.body.date), row = db.prepare('SELECT off, blocked_dates FROM pandits WHERE id=?').get(pid(req)), off = j(row.off, []);
  /* If the date is in the structured blocked list, removing it there takes priority
     so the legacy toggle can un-block what it previously blocked. */
  const blocked = j(row.blocked_dates, []);
  const bi = blocked.findIndex((b) => b && b.date === date);
  if (bi > -1) { blocked.splice(bi, 1); db.prepare('UPDATE pandits SET blocked_dates=? WHERE id=?').run(JSON.stringify(blocked), pid(req)); }
  const i = off.indexOf(date);
  if (i > -1) off.splice(i, 1); else off.push(date);
  db.prepare('UPDATE pandits SET off=? WHERE id=?').run(JSON.stringify(off), pid(req));
  res.json({ ok: true });
});

/* --- Centralized availability calendar (Phase 3): rules + per-date overrides --- */
router.get('/calendar', (req, res) => {
  const row = db.prepare('SELECT * FROM pandits WHERE id=?').get(pid(req));
  res.json({ calendar: require('../services/availability').configOf(row) });
});
router.put('/calendar', (req, res) => {
  const b = req.body || {};
  const AV = require('../services/availability');
  const cur = db.prepare('SELECT * FROM pandits WHERE id=?').get(pid(req));
  const weekly = (Array.isArray(b.weeklyOff) ? b.weeklyOff : []).map((n) => v.int(n, 'Weekday', { min: 0, max: 6 }));
  const slots = (Array.isArray(b.slots) ? b.slots : []).filter((s) => P.SLOTS.includes(s));
  const radius = b.radiusKm === null || b.radiusKm === undefined || b.radiusKm === '' ? null : v.int(b.radiusKm, 'Radius', { min: 1, max: 500 });
  const base = (() => {
    if (b.baseLat != null && b.baseLon != null) return { lat: Number(b.baseLat), lon: Number(b.baseLon) };
    if (b.baseCity) { const pl = AV.resolvePlace(b.baseCity); if (pl) return { lat: pl.lat, lon: pl.lon }; }
    return null;
  })();
  if (radius != null && !base && cur.base_lat == null && !AV.resolvePlace(cur.city)) throw bad('Set a base location (or a recognizable city) before enabling the service radius');
  db.prepare(`UPDATE pandits SET weekly_off=?, slots=?, radius_km=?, base_lat=?, base_lon=?,
    online_enabled=?, temple_enabled=? WHERE id=?`)
    .run(JSON.stringify([...new Set(weekly)].sort()), JSON.stringify(slots), radius,
         base ? base.lat : cur.base_lat, base ? base.lon : cur.base_lon,
         b.onlineEnabled === undefined ? cur.online_enabled : (b.onlineEnabled ? 1 : 0),
         b.templeEnabled === undefined ? cur.temple_enabled : (b.templeEnabled ? 1 : 0), pid(req));
  res.json({ calendar: AV.configOf(db.prepare('SELECT * FROM pandits WHERE id=?').get(pid(req))) });
});
router.post('/calendar/dates', (req, res) => {
  const b = req.body || {}, date = v.date(b.date);
  const AV = require('../services/availability');
  const row = db.prepare('SELECT off, holidays, blocked_dates FROM pandits WHERE id=?').get(pid(req));
  const holidays = j(row.holidays, []), blocked = j(row.blocked_dates, []);
  const kind = v.oneOf(b.kind || 'holiday', ['holiday', 'blocked'], 'Kind');
  if (kind === 'holiday') {
    const i = holidays.indexOf(date);
    if (i > -1) holidays.splice(i, 1); else holidays.push(date);
  } else {
    const i = blocked.findIndex((x) => x && x.date === date);
    if (i > -1) blocked.splice(i, 1);
    else blocked.push({ date, reason: String(b.reason || '').slice(0, 120) });
  }
  db.prepare('UPDATE pandits SET holidays=?, blocked_dates=? WHERE id=?').run(JSON.stringify(holidays), JSON.stringify(blocked), pid(req));
  res.json({ calendar: AV.configOf(db.prepare('SELECT * FROM pandits WHERE id=?').get(pid(req))) });
});
/* Why can't this pandit take a given booking? Used by the portal calendar UX. */
router.get('/calendar/why', (req, res) => {
  const date = v.date(req.query.date), slot = req.query.slot && P.SLOTS.includes(req.query.slot) ? req.query.slot : null;
  const row = db.prepare('SELECT * FROM pandits WHERE id=?').get(pid(req));
  res.json({ verdict: require('../services/availability').check(row, date, slot, { mode: req.query.mode }) });
});
router.patch('/profile', (req, res) => {
  const b = req.body, spec = list(b.spec).filter((s) => pujaIds().includes(s));
  if (!spec.length) throw bad('Select at least one puja');
  db.prepare('UPDATE pandits SET city=?, exp=?, langs=?, bio=?, spec=?, avail=? WHERE id=?')
    .run(CITIES.includes(b.city) ? b.city : 'Delhi NCR', v.int(b.exp, 'Experience', { min: 0, max: 70 }), JSON.stringify(list(b.langs).slice(0, 10)), v.str(b.bio, 'About', { optional: true, max: 600 }), JSON.stringify(spec), b.avail ? 1 : 0, pid(req));
  res.json({ ok: true });
});
router.post('/feature', (req, res) => {
  if ((process.env.PAYMENT_MODE || 'mock') !== 'mock') return res.status(501).json({ error: 'Featured-listing billing is not wired to the gateway yet. See README.' });
  db.prepare('UPDATE pandits SET featured=1 WHERE id=?').run(pid(req));
  res.json({ ok: true });
});

/* --- My puja photos: upload only for own assigned bookings; pending until the
   admin approves. Pandit sees their own uploads with approval status. --- */
const M = require('../services/pujaMedia');
const { wrap } = require('../lib/util');
router.post('/media', upload.media.array('media', 8), upload.verifyMagic(), wrap(async (req, res) => {
  if (!req.files || !req.files.length) throw bad('Attach at least one photo');
  if (!String(req.body.altText || '').trim()) throw bad('Describe the photo (alt text is required)');
  const media = M.panditUpload({ pid: pid(req), uid: req.auth.uid, bookingId: req.body.bookingId, files: req.files, altText: req.body.altText });
  /* Variants are generated inline (awaited) so the client never sees rows whose
     derived artifacts may still be in flight — a former background-repair race. */
  for (const m of media) await require('../services/mediaVariants').ensureVariants(M.row(m.id));
  res.status(201).json({ media });
}));
router.get('/media', (req, res) => res.json({ media: M.mineForPandit(pid(req)) }));
router.delete('/media/:id', (req, res) => res.json(M.remove({ uid: req.auth.uid, role: 'pandit', pid: pid(req), id: req.params.id })));
module.exports = router;

File: backend-python/app/routers/pandit.py
"""Pandit routes — port of server/routes/pandit.js booking actions, availability,
profile and featured listing (media endpoints already live in routers/media.py;
pandit registration arrives with KYC uploads)."""
import json

from fastapi import APIRouter, Depends, UploadFile, File
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..db import get_db
from ..models import Pandit, Puja
from ..security import require_role
from ..serialize import booking as s_booking
from ..services import bookings as B
from ..services.availability import check as av_check, config_of, resolve_place
from ..util import bad, j, v_date, v_int, v_one_of, v_str
from ..pricing import SLOTS

router = APIRouter(prefix="/api/pandit", tags=["pandit"])
pandit_dep = require_role("pandit")

CITIES = ["Delhi NCR", "Mumbai", "Bengaluru", "Pune", "Jaipur", "Lucknow",
          "Varanasi", "Ahmedabad", "Chennai", "Hyderabad"]


def _list(x) -> list[str]:
    if isinstance(x, list):
        return [str(s).strip() for s in x if str(s).strip()]
    return [s.strip() for s in str(x or "").split(",") if s.strip()]


async def _verified(db: AsyncSession, pid: str) -> None:
    p = await db.get(Pandit, pid)
    if not p or p.status != "verified":
        raise bad("Your KYC is not verified yet")


@router.post("/bookings/{booking_id}/complete")
async def pandit_complete(booking_id: str,
                          media: list[UploadFile] | None = File(None),
                          auth: dict = Depends(pandit_dep),
                          db: AsyncSession = Depends(get_db)):
    """Declared BEFORE the generic /{action} route so FastAPI matches it first
    (Node used a regex route that excluded 'complete'; here order does it)."""
    await _verified(db, auth["pid"])
    from ..config import get_settings
    from ..util import rid, verify_upload
    from pathlib import Path
    urls = []
    for f in (media or [])[:8]:
        data = await f.read()
        real = verify_upload(data, f.content_type or "")
        name = rid(8) + {"image/jpeg": ".jpg", "image/png": ".png",
                         "image/webp": ".webp"}[real]
        d = Path(get_settings().upload_dir) / "media"
        d.mkdir(parents=True, exist_ok=True)
        (d / name).write_bytes(data)
        urls.append("/media/" + name)
    row = await B.pandit_act(db, auth["pid"], booking_id, "complete", urls)
    return {"booking": s_booking(row)}


@router.post("/bookings/{booking_id}/{action}")
async def pandit_booking_action(booking_id: str, action: str,
                                auth: dict = Depends(pandit_dep),
                                db: AsyncSession = Depends(get_db)):
    if action not in ("accept", "reject", "start"):
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Not found")
    await _verified(db, auth["pid"])
    row = await B.pandit_act(db, auth["pid"], booking_id, action)
    return {"booking": s_booking(row)}


@router.post("/availability")
async def toggle_availability(body: dict, auth: dict = Depends(pandit_dep),
                              db: AsyncSession = Depends(get_db)):
    date = v_date((body or {}).get("date"))
    p = await db.get(Pandit, auth["pid"])
    off = j(p.off, [])
    # A structured blocked entry takes priority: the legacy toggle un-blocks it.
    blocked = j(p.blocked_dates, [])
    bi = next((i for i, b in enumerate(blocked) if isinstance(b, dict) and b.get("date") == date), None)
    if bi is not None:
        blocked.pop(bi)
        p.blocked_dates = json.dumps(blocked)
    if date in off:
        off.remove(date)
    else:
        off.append(date)
    p.off = json.dumps(off)
    await db.flush()
    return {"ok": True}


# --- Centralized availability calendar (Phase 3) ---------------------------------
@router.get("/calendar")
async def get_calendar(auth: dict = Depends(pandit_dep), db: AsyncSession = Depends(get_db)):
    p = await db.get(Pandit, auth["pid"])
    return {"calendar": config_of(p)}


@router.put("/calendar")
async def put_calendar(body: dict, auth: dict = Depends(pandit_dep),
                       db: AsyncSession = Depends(get_db)):
    b = body or {}
    p = await db.get(Pandit, auth["pid"])
    weekly = sorted({v_int(n, "Weekday", min_val=0, max_val=6)
                     for n in (b.get("weeklyOff") or []) if isinstance(n, int)})
    slots = [s for s in (b.get("slots") or []) if s in SLOTS]
    radius = b.get("radiusKm")
    radius = None if radius in (None, "") else v_int(radius, "Radius", min_val=1, max_val=500)
    base = None
    if b.get("baseLat") is not None and b.get("baseLon") is not None:
        base = (float(b["baseLat"]), float(b["baseLon"]))
    elif b.get("baseCity"):
        pl = await resolve_place(db, b["baseCity"])
        base = (pl.lat, pl.lon) if pl else None
    if radius is not None and base is None and p.base_lat is None and not await resolve_place(db, p.city):
        raise bad("Set a base location (or a recognizable city) before enabling the service radius")
    p.weekly_off = json.dumps(weekly)
    p.slots = json.dumps(slots)
    p.radius_km = radius
    if base is not None:
        p.base_lat, p.base_lon = base
    if b.get("onlineEnabled") is not None:
        p.online_enabled = 1 if b["onlineEnabled"] else 0
    if b.get("templeEnabled") is not None:
        p.temple_enabled = 1 if b["templeEnabled"] else 0
    await db.flush()
    await db.refresh(p)
    return {"calendar": config_of(p)}


@router.post("/calendar/dates")
async def calendar_dates(body: dict, auth: dict = Depends(pandit_dep),
                         db: AsyncSession = Depends(get_db)):
    b = body or {}
    date = v_date(b.get("date"))
    kind = v_one_of(b.get("kind") or "holiday", ["holiday", "blocked"], "Kind")
    p = await db.get(Pandit, auth["pid"])
    holidays = j(p.holidays, [])
    blocked = j(p.blocked_dates, [])
    if kind == "holiday":
        if date in holidays:
            holidays.remove(date)
        else:
            holidays.append(date)
    else:
        idx = next((i for i, x in enumerate(blocked)
                    if isinstance(x, dict) and x.get("date") == date), None)
        if idx is not None:
            blocked.pop(idx)
        else:
            blocked.append({"date": date, "reason": str(b.get("reason") or "")[:120]})
    p.holidays = json.dumps(holidays)
    p.blocked_dates = json.dumps(blocked)
    await db.flush()
    await db.refresh(p)
    return {"calendar": config_of(p)}


@router.get("/calendar/why")
async def calendar_why(date: str, slot: str | None = None, mode: str | None = None,
                       auth: dict = Depends(pandit_dep), db: AsyncSession = Depends(get_db)):
    date_v = v_date(date)
    slot_v = slot if slot in SLOTS else None
    p = await db.get(Pandit, auth["pid"])
    return {"verdict": await av_check(db, p, date_v, slot_v, mode=mode)}


@router.patch("/profile")
async def update_profile(body: dict, auth: dict = Depends(pandit_dep),
                         db: AsyncSession = Depends(get_db)):
    b = body or {}
    puja_ids = set((await db.execute(select(Puja.id))).scalars().all())
    spec = [s for s in _list(b.get("spec")) if s in puja_ids]
    if not spec:
        raise bad("Select at least one puja")
    p = await db.get(Pandit, auth["pid"])
    p.city = b.get("city") if b.get("city") in CITIES else "Delhi NCR"
    p.exp = v_int(b.get("exp"), "Experience", min_val=0, max_val=70)
    p.langs = json.dumps(_list(b.get("langs"))[:10])
    p.bio = v_str(b.get("bio"), "About", optional=True, max_len=600)
    p.spec = json.dumps(spec)
    p.avail = 1 if b.get("avail") else 0
    await db.flush()
    return {"ok": True}


@router.post("/feature")
async def get_featured(auth: dict = Depends(pandit_dep),
                       db: AsyncSession = Depends(get_db)):
    import os
    if (os.environ.get("PAYMENT_MODE") or "mock") != "mock":
        from ..util import http_error
        raise http_error(501, "Featured-listing billing is not wired to the gateway yet. See README.")
    p = await db.get(Pandit, auth["pid"])
    p.featured = 1
    await db.flush()
    return {"ok": True}
